frontend/server/controllers/ContestController.php

```diff
@@ -2345,7 +2368,55 @@ public static function apiUpdate(Request $r) {
     }
 
     /**
-     * This function reviews changes in penalty type and admission mode
+     * Update Contest end time for an identity when window_length
+     * option is turned on
+     *
+     * @param Request $r
+     * @return array
+     * @throws InvalidDatabaseOperationException
+     * @throws NotFoundException
+     */
+    public static function apiUpdateEndTimeForIdentity(Request $r) {
+        if (OMEGAUP_LOCKDOWN) {
+            throw new ForbiddenAccessException('lockdown');
+        }
+
+        self::authenticateRequest($r);
+        self::validateBasicDetails($r);
+
+        if (is_null($r['contest'])) {
+            throw new NotFoundException('contestNotFound');
+        }
+
+        if (!Authorization::isContestAdmin($r['current_identity_id'], $r['contest'])) {
+            throw new ForbiddenAccessException();
+        }
+
+        Validators::validateStringNonEmpty($r['username'], 'username');
+        Validators::validateNumber($r['end_time'], 'end_time');
+
+        try {
+            $updatedEntries = ProblemsetIdentitiesDAO::updateEndTimeForIdentity(
+                $r['problemset']->problemset_id,
+                $r['username'],
+                gmdate('Y-m-d H:i:s', $r['end_time'])
+            );
+        } catch (Exception $e) {
+            // Operation failed in the data layer
+            throw new InvalidDatabaseOperationException($e);
+        }
+
+        if (!$updatedEntries) {
+            throw new NotFoundException('ProblemsetIdentityNotFound');
```

Comment:
esta cadena no existe D: puedes crearla? y de paso

```suggestion
            throw new NotFoundException('problemsetIdentityNotFound');
```

Reply:
Listo